db/merge_helper.cc

```diff
@@ -57,120 +58,297 @@ MergeHelper::MergeHelper(Env* env, const Comparator* user_comparator,
-      if (result_operand != nullptr) {
-        *result_operand = tmp_result_operand;
-      } else {
-        result->assign(tmp_result_operand.data(), tmp_result_operand.size());
-      }
-    } else if (result_operand) {
-      *result_operand = Slice(nullptr, 0);
-    }
+    success = merge_operator->FullMergeV3(merge_in, &merge_out);
 
     RecordTick(statistics, MERGE_OPERATION_TOTAL_TIME,
                statistics ? timer.ElapsedNanos() : 0);
   }
 
-  if (op_failure_scope != nullptr) {
-    *op_failure_scope = merge_out.op_failure_scope;
-    // Apply default per merge_operator.h
-    if (*op_failure_scope == MergeOperator::OpFailureScope::kDefault) {
-      *op_failure_scope = MergeOperator::OpFailureScope::kTryMerge;
-    }
-  }
-
   if (!success) {
     RecordTick(statistics, NUMBER_MERGE_FAILURES);
+
+    if (op_failure_scope) {
+      *op_failure_scope = merge_out.op_failure_scope;
+      // Apply default per merge_operator.h
+      if (*op_failure_scope == MergeOperator::OpFailureScope::kDefault) {
+        *op_failure_scope = MergeOperator::OpFailureScope::kTryMerge;
+      }
+    }
+
     return Status::Corruption(Status::SubCode::kMergeOperatorFailed);
   }
 
-  return Status::OK();
+  return std::visit(std::forward<Visitor>(visitor),
+                    std::move(merge_out.new_value));
 }
 
-Status MergeHelper::TimedFullMergeWithEntity(
-    const MergeOperator* merge_operator, const Slice& key, Slice base_entity,
-    const std::vector<Slice>& operands, std::string* result, Logger* logger,
+Status MergeHelper::TimedFullMergeImpl(
+    const MergeOperator* merge_operator, const Slice& key,
+    MergeOperator::MergeOperationInputV3::ExistingValue&& existing_value,
+    const std::vector<Slice>& operands, Logger* logger, Statistics* statistics,
+    SystemClock* clock, bool update_num_ops_stats, std::string* result,
+    Slice* result_operand, ValueType* result_type,
+    MergeOperator::OpFailureScope* op_failure_scope) {
+  assert(result);
+  assert(result_type);
+
+  auto visitor = overload{
+      [&](std::string&& new_value) -> Status {
+        *result_type = kTypeValue;
+
+        if (result_operand) {
+          *result_operand = Slice(nullptr, 0);
+        }
+
+        *result = std::move(new_value);
+
+        return Status::OK();
+      },
+      [&](MergeOperator::MergeOperationOutputV3::NewColumns&& new_columns)
+          -> Status {
+        *result_type = kTypeWideColumnEntity;
+
+        if (result_operand) {
+          *result_operand = Slice(nullptr, 0);
+        }
+
+        result->clear();
+
+        WideColumns sorted_columns;
+        sorted_columns.reserve(new_columns.size());
+
+        for (const auto& column : new_columns) {
+          sorted_columns.emplace_back(column.first, column.second);
+        }
+
+        WideColumnsHelper::SortColumns(sorted_columns);
+
+        return WideColumnSerialization::Serialize(sorted_columns, *result);
+      },
+      [&](Slice&& operand) {
```

Comment:
Still learning the C++ syntax here... I guess `Status` being returned when the new_value is `Slice` can be implicitly determined?  

Basically I was expecting`[&](Slice&& operand) -> Status {` in this line

Reply:
Yes, the compiler can deduce the return type in this case; will add the trailing `-> Status` for consistency though. Thanks for catching this!